### ISSUE TITLE
VanillaChronicleHash.allocateTier IllegalStateException

### ISSUE THREAD

**Issue by mstrap** [OpenHFT/Chronicle-Map#146]
We are seeing following exception rather frequently:

```
java.lang.IllegalStateException
	at net.openhft.chronicle.hash.impl.VanillaChronicleHash.allocateTier(VanillaChronicleHash.java:786)
	at net.openhft.chronicle.map.impl.CompiledMapQueryContext.nextTier(CompiledMapQueryContext.java:3126)
	at net.openhft.chronicle.map.impl.CompiledMapQueryContext.alloc(CompiledMapQueryContext.java:3487)
	at net.openhft.chronicle.map.impl.CompiledMapQueryContext.initEntryAndKey(CompiledMapQueryContext.java:3505)
	at net.openhft.chronicle.map.impl.CompiledMapQueryContext.putEntry(CompiledMapQueryContext.java:3998)
	at net.openhft.chronicle.map.impl.CompiledMapQueryContext.doInsert(CompiledMapQueryContext.java:4187)
	at net.openhft.chronicle.map.MapEntryOperations.insert(MapEntryOperations.java:157)
	at net.openhft.chronicle.map.impl.CompiledMapQueryContext.insert(CompiledMapQueryContext.java:4110)
	at net.openhft.chronicle.map.MapMethods.put(MapMethods.java:89)
	at net.openhft.chronicle.map.VanillaChronicleMap.put(VanillaChronicleMap.java:719)
```

I think I got `builder.entries()`, `averageKeySize()` and `averageValueSize()`. Not sure about ` maxBloatFactor()`. Also I can't rule out "keys, inserted into the ChronicleHash, are distributed suspiciously bad".

How should one typically approach/address this problem? Is there a workaround?

**Comment by leventov** [comment]
It should post "Attempt to allocate #X, extra segment tier, Y is maximum". Could you post your X, Y, also the number you configured in `builder.entries()`.

> Also I can't rule out "keys, inserted into the ChronicleHash, are distributed suspiciously bad".

To check this theory, please call `MapDiagnostics.printMapStats(map)` when you get an `IllegalStateException`. (Strangely I don't see this in Javadocs, but at least you could replicate the code easily, it uses only public interfaces: https://github.com/OpenHFT/Chronicle-Map/blob/master/src/main/java/net/openhft/chronicle/map/MapDiagnostics.java#L33)

**Comment by mstrap** [comment]
Thanks. Unfortunately the errors are created on our customer machines and I don't have the request logging available. I'm closing this until it becomes an issue again.

**Comment by hft-team-city** [comment]
Released in Chronicle-Map-3.14.6 (https://github.com/OpenHFT/Chronicle-Map/releases/chronicle-map-3.14.6), BOM-1.15.271 (https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-1.15.271)

**Comment by hft-team-city** [comment]
Released in Chronicle-Map-3.14.6 (https://github.com/OpenHFT/Chronicle-Map/releases/chronicle-map-3.14.6), BOM-1.15.272 (https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-1.15.272)

**Comment by hft-team-city** [comment]
Released in Chronicle-Map-3.14.6 (https://github.com/OpenHFT/Chronicle-Map/releases/chronicle-map-3.14.6), BOM-1.15.274 (https://github.com/OpenHFT/OpenHFT/releases/chronicle-bom-1.15.274)
